### PR TITLE
Refactor Mask CTC non-autoregressive ASR

### PR DESCRIPTION
**Pull request by YosukeHiguchi**
This PR is for refactoring #2070 .
- [x] modularize mask ctc using existing transformer model
- [x] add documentation
- [x] add config files for CSJ, Voxforge, Tedlium2

### CODE REVIEW

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    @staticmethod
+    def add_arguments(parser):
+        """Add arguments."""
+        E2ETransformer.add_arguments(parser)
+        E2E.add_maskctc_arguments(parser)
+
+        return parser
+
+    @staticmethod
+    def add_maskctc_arguments(parser):
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:42]: No additional argument now?

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        :rtype: float
+        """
+        # 1. forward encoder
+        xs_pad = xs_pad[:, : max(ilens)]  # for data parallel
+        src_mask = make_non_pad_mask(ilens.tolist()).to(xs_pad.device).unsqueeze(-2)
+        hs_pad, hs_mask = self.encoder(xs_pad, src_mask)
+        self.hs_pad = hs_pad
+
+        # 2. forward decoder
+        if self.decoder is not None:
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:85]: You can ignore pure CTC case by checking whether `mtlalpha > 0`.

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        n2s = num2str(char_list, self.mask_token)
+
+        self.eval()
+        h = self.encode(x).unsqueeze(0)
+
+        ctc_probs, ctc_ids = torch.exp(self.ctc.log_softmax(h)).max(dim=-1)
+        y_hat = torch.stack([x[0] for x in groupby(ctc_ids[0])])
+        y_idx = torch.nonzero(y_hat != 0).squeeze(-1)
+
+        probs_hat = []
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:169]: Please add comments for these lines.

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        if not mask_num == 0:
+            K = recog_args.maskctc_n_iterations
+            num_iter = K if mask_num >= K and K > 0 else mask_num
+
+            for t in range(num_iter - 1):
+                pred, _ = self.decoder(
+                    y_in, (y_in != self.ignore_id).unsqueeze(-2), h, None
+                )
+                pred_sc, pred_id = pred[0][mask_idx].max(dim=-1)
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:198]: I recommend replacing `pred_sc` with `pred_score` or `logits`.

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        probs_hat = torch.from_numpy(numpy.array(probs_hat))
+
+        p_thres = recog_args.maskctc_probability_threshold
+        mask_idx = torch.nonzero(probs_hat[y_idx] < p_thres).squeeze(-1)
+        confident_idx = torch.nonzero(probs_hat[y_idx] >= p_thres).squeeze(-1)
+        mask_num = len(mask_idx)
+
+        y_in = torch.zeros(1, len(y_idx) + 1, dtype=torch.long) + self.mask_token
+        y_in[0][confident_idx] = y_hat[y_idx][confident_idx]
+        y_in[0][-1] = self.eos
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:186]: Is this necessary?

**Comment by sw005320** [comment]
Could you make it work without preparing a special recipe (`local/run_maskctc.sh`)?
I think this can be run by only providing a config file.
Also, could you prepare a config file for the other recipe that you show some effectiveness in your paper?

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py (hunk 1 above)
Comment (YosukeHiguchi) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py:42]: No for now, should I remove this method?

Comment (sw005320) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py:42]: Will you have something in the near future?
I'm fine to keep it for such a future preparation for now.

Comment (YosukeHiguchi) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py:42]: ok, I'll keep it for the future development.

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py (hunk 2 above)
Comment (hirofumi0810) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py]: @YosukeHiguchi Sorry for my ambiguous comments. I mean you can just check `mtlalpha` at once when building the model, not in the forward pass.

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        cer_ctc = None
+        if self.mtlalpha == 0.0:
+            loss_ctc = None
+        else:
+            batch_size = xs_pad.size(0)
+            hs_len = hs_mask.view(batch_size, -1).sum(1)
+            loss_ctc = self.ctc(hs_pad.view(batch_size, -1, self.adim), hs_len, ys_pad)
+            if self.error_calculator is not None:
+                ys_hat = self.ctc.argmax(hs_pad.view(batch_size, -1, self.adim)).data
+                cer_ctc = self.error_calculator(ys_hat.cpu(), ys_pad.cpu(), is_ctc=True)
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:108]:
```suggestion
        loss_ctc, cer_ctc = None, None
        if self.mtlalpha > 0:
            batch_size = xs_pad.size(0)
            hs_len = hs_mask.view(batch_size, -1).sum(1)
            loss_ctc = self.ctc(hs_pad.view(batch_size, -1, self.adim), hs_len, ys_pad)
            if self.error_calculator is not None:
                ys_hat = self.ctc.argmax(hs_pad.view(batch_size, -1, self.adim)).data
                cer_ctc = self.error_calculator(ys_hat.cpu(), ys_pad.cpu(), is_ctc=True)
```

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+        # iterative decoding
+        if not mask_num == 0:
+            K = recog_args.maskctc_n_iterations
+            num_iter = K if mask_num >= K and K > 0 else mask_num
+
+            for t in range(num_iter - 1):
+                pred, _ = self.decoder(
+                    y_in, (y_in != self.ignore_id).unsqueeze(-2), h, None
+                )
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:193]:
```suggestion
                pred, _ = self.decoder(
                    y_in, None, h, None
                )
```

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                cand = torch.topk(pred_score, mask_num // num_iter, -1)[1]
+                y_in[0][mask_idx[cand]] = pred_id[cand]
+                mask_idx = torch.nonzero(y_in[0] == self.mask_token).squeeze(-1)
+
+                logging.info("msk:{}".format(n2s(y_in[0].tolist())))
+
+            # predict leftover masks (|masks| < mask_num // num_iter)
+            pred, pred_mask = self.decoder(
+                y_in, (y_in != self.ignore_id).unsqueeze(-2), h, None
+            )
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_maskctc.py:204]:
```suggestion
            pred, pred_mask = self.decoder(
                y_in, None, h, None
            )
```

File: espnet/nets/pytorch_backend/e2e_asr_maskctc.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        xs_pad = xs_pad[:, : max(ilens)]  # for data parallel
+        src_mask = make_non_pad_mask(ilens.tolist()).to(xs_pad.device).unsqueeze(-2)
+        hs_pad, hs_mask = self.encoder(xs_pad, src_mask)
+        self.hs_pad = hs_pad
+
+        # 2. forward decoder
+        ys_in_pad, ys_out_pad = mask_uniform(
+            ys_pad, self.mask_token, self.eos, self.ignore_id
+        )
+        ys_mask = (ys_in_pad != self.ignore_id).unsqueeze(-2)
```
[the remark this answers is not in the document]
Comment (xingchensong) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py:88]: > 

I think this may not be an error? `ys_mask` is used for masking pad-token in decoder's self-attention thus should be calculated from `ys_in_pad`. 
In fact, the `ys_mask` calculated using `ys_out_pad` is meaningful only when calculating accuracy or loss, this has been done inside function `th_accuracy` and forwardpass of  `self.criterion`.

[the remark this answers is not in the document]
Comment (hirofumi0810) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py:88]: My point is that `ys_in_pad` is padded with `eos`. So the expected padding indices are not padded accutually.

[the remark this answers is not in the document]
Comment (YosukeHiguchi) [reply espnet/nets/pytorch_backend/e2e_asr_maskctc.py]: Thank you for the comments. This way of padding is not correct as hirofumi pointed out. I'll rerun the experiments to see if this bug has effects on mask ctc's performance.

**Review by sw005320 — APPROVED** [review]
LGTM

**Comment by sw005320** [comment]
@hirofumi0810, could you review it again?
I think it is already in a good shape, and I'll merge this soon after your review.

**Comment by hirofumi0810** [comment]
LGTM. I might update some scripts to make them compatible with NAR E2E-ST.

**Review by hirofumi0810 — APPROVED** [review]

**Comment by sw005320** [comment]
Thanks a lot, @YosukeHiguchi and @hirofumi0810 !
